src/openforms/prefill/__init__.py

```diff
@@ -76,6 +78,75 @@ def apply_prefill(configuration: JSONObject, submission: "Submission", register=
     fields = _extract_prefill_fields(configuration)
     grouped_fields = _group_prefills_by_plugin(fields)
 
+    results = _fetch_prefill_values_cached(grouped_fields, submission, register)
+
+    # finally, ensure the ``defaultValue`` is set based on prefill results
+    config_copy = deepcopy(configuration)
+    _set_default_values(config_copy, results)
+    return config_copy
+
+
+def _fetch_prefill_values_cached(
+    grouped_fields: Dict[str, list], submission: "Submission", register
+) -> Dict[str, Dict[str, Any]]:
+    """
+    Wraps _fetch_prefill_values() with similar signature but caches on submission.prefill_data
+
+    As complication the prefill is invoked per FormStep,
+        but different steps could have same plugin or even repeating prefills from same plugins
+
+    This means the cache may have values we're currently not interested in,
+        but want to keep for a different request
+    """
+    results: Dict[str, Dict[str, Any]] = dict()
+
+    cached = submission.prefill_data
+    cache_dirty = False
+
+    # grab what we want from cache and collect misses
+    fetch_fields = defaultdict(list)
+    for plugin_id, fields in grouped_fields.items():
+        for field in fields:
+            path = Path(plugin_id, field)
+            try:
+                assign(results, path, glom(cached, path), missing=dict)
+            except GlomError:
+                fetch_fields[plugin_id].append(field)
+
+    # fetch missing values
+    if fetch_fields:
+        fetch_results = _fetch_prefill_values(fetch_fields, submission, register)
+
+        # copy to result and update cache
+        for plugin_id, values in fetch_results.items():
+            for field, value in values.items():
+                path = Path(plugin_id, field)
+                assign(results, path, value, missing=dict)
+                assign(cached, path, value, missing=dict)
+                cache_dirty = True
+
+        # set None for fields not returned by plugin to block endless re-requesting
+        for plugin_id, fields in fetch_fields.items():
+            for field in fields:
+                path = Path(plugin_id, field)
+                try:
+                    glom(results, path)
+                except GlomError:
+                    assign(results, path, None, missing=dict)
+                    assign(cached, path, None, missing=dict)
+                    cache_dirty = True
```

Comment:
This is too naive I think, the type of the data should match the type of the component.

Looking at some Formio configurations, I can see `defaultValue: ""` for text-based components. The pre-fill machinery with a pre-fill match would change that into `defaultValue: null`, which is not correct.

I'm okay with storing `None` in the DB (JSONField) as that is semantically correct, but we need to process that when setting the values so that the `defaultValue` is untouched if the result is `None`.

Possibly the implementation is correct, and the tests are set up wrong :-) so please check what happens after you fix the tests first.